### DISCUSSION TITLE
Would Dace be a reasonable place to start for HLS for use with a KV260?

### DISCUSSION THREAD

**Discussion by rog77** [spcl/dace#1652] [Q&A] [answered]
I am interested in using this software for generating kv260 bitstreams from Python code, and then using them as overlays in PYNQ - That is to say I hope to be able to run python programs on the Arm/PS side of the chip, and have it call/pass data to be executed on the PL/FPGA side that was generated with DaCe.

I am new to HLS but OK with python, and there seem to be loads of options, and I want to be sure that I an going in the right direction before I start in earnest.

For example I might want a matmul or matrix/vector dot product written in DaCe and compiled/synthesised ahead of time as a bitstream/overlay *specifically* for the xilinx kv260, and then to be able to load/call it from PYNQ on the device at runtime.

Does this sort of thing seem a reasonable aspiration with DaCe as it stands right now? Are there any pitfalls I should be aware of? I already have an algo in python/numpy thst I wish to accelerate, and really don't want to have to learn C++/HLS if this library will do the job without too much of a performance/utilisation hit.

Thanks!

**Comment by tbennun** [comment, marked as answer]
This sounds like the standard way we use DaCe to run FPGA codes (invoking a compiled bitstream from Python and letting it handle all the HLS and data copying). 

I would suggest to read the relevant documentation entries: [Writing DaCe programs in Python](https://spcldace.readthedocs.io/en/latest/frontend/daceprograms.html), [FPGA GEMM code sample written in numpy](https://github.com/spcl/dace/blob/master/tests/npbench/polybench/gemm_npbench_test.py#L59), [FPGA optimization best practices](https://spcldace.readthedocs.io/en/latest/optimization/fpga.html), and if you wish to go low-level, [how FPGA HLS code generation works](https://spcldace.readthedocs.io/en/latest/codegen/codegen.html#fpga-code-generation) (the whole page is a useful read in my opinion).

For pitfalls, remember to change the `.dace.conf` file to the appropriate hardware string for the KV260 platform: https://spcldace.readthedocs.io/en/latest/source/config_schema.html#envvar-compiler.xilinx.platform
and change the compilation mode to `hardware` rather than software simulation: https://spcldace.readthedocs.io/en/latest/source/config_schema.html#envvar-compiler.xilinx.mode

**Reply by rog77** [reply]
This is very useful, thank you! If it's OK I will keep this thread open in case I have further related questions/issues.
